qa: Facilitate debugging bitcoind inside functional tests

hodlinator opened the pull request:
Facilitates debugging specific instances of bitcoind *in the context of Python tests, inside of an editor/IDE/debugger*.

Makes bitcoind spin during startup, waiting for a debugger to be attached, so that breakpoints during process init can be debugged.

### Example usage

1. Generate build files with `-DENABLE_WAIT_FOR_DEBUGGER=ON` & build.
   
   ```
   cmake -B build -DCMAKE_BUILD_TYPE=Debug -DENABLE_WAIT_FOR_DEBUGGER=ON && cmake --build build -j$(nproc)
   ```

2. Set breakpoints in the C++ code in your editor/IDE/debugger.

3. Start a Python functional test with a debugger command such as:
   - LLDB inside an already started Tmux pane (append ` -o continue` to the LLDB args to have the debugger continue automatically):
     ```
     ./build/test/functional/feature_loadblock.py --debug_runs 2 --debug_cmd "tmux split-window -h lldb -p \$PID\$"
     ```
   - Sublime Text Debugger, allows a running session to attach:
     ```
     ./build/test/functional/feature_loadblock.py --debug_runs 2 --debug_cmd "subl --command \"debugger {\\\"action\\\": \\\"start\\\", \\\"configuration\\\": {\\\"name\\\": \\\"foo\\\", \\\"type\\\": \\\"lldb\\\", \\\"request\\\": \\\"attach\\\", \\\"program\\\": \\\"dummyprogram\\\", \\\"pid\\\": \\\"\$PID\$\\\"}}\""
     ```
   - Don't specify `--debug_cmd ...\$PID\$...`, instead have the test framework log which bitcoind PID you should manually attach to:
     ```
     ./build/test/functional/feature_loadblock.py --debug_runs 2
     ```

5. Experience how debugger hits breakpoints as the functional test executes RPCs.

Using `-debug_runs <RUN_INDEX>` often requires doing a dry-run of a functional test with `-ldebug` and examining the log to see which run index is the one we want to debug:
```
2025-...Z TestFramework.node0 (DEBUG): bitcoind started (run #1, node #0), waiting for RPC to come up
``` 

### Alternative usage

Tweak functional test code to specify `wait_for_debugger` when (re)starting a node:

```diff
-        self.restart_node(0)
+        self.restart_node(0, wait_for_debugger=True)
```
While it can be messy to modify files while moving between commits, and cumbersome inside of loops, it might be more straightforward than `-debug_runs <RUN_INDEX>` in a given case.

### *Needed before merge*

- [ ] Testing on Mac to confirm -waitfordebugger on the C++ side behaves as expected.

fanquake commented:
> Makes bitcoind spin during startup, waiting for a debugger to be attached.

Can you explain more? If you want to run under a debugger, then why not just start `bitcoind` under a debugger; why do we need to add this option?

hodlinator commented:
> Can you explain more? If you want to run under a debugger, then why not just start `bitcoind` under a debugger; why do we need to add this option?

Clarified PR description text and added example usage. But I'm curious to hear other ways of doing the same thing if there are any.

hodlinator commented on src/bitcoind.cpp:
```diff
@@ -254,6 +263,51 @@ static bool AppInit(NodeContext& node)
 MAIN_FUNCTION
 {
+    for (int i = 0; i < argc; ++i) {
+        if (strcmp(argv[i], "-waitfordebugger") == 0) {
```
Updated PR description to clarify intended usage.

One could possibly launch *bitcoind* as a parameter to *lldb*/*gdb*, but I don't know if there is a way of doing that if one also wants to have the debugging view with stepping etc inside of the editor/IDE.

hodlinator commented:
> I think if you want to debug the start-up sequence specifically, it may be easier to do it outside the tests? Then anything else, if needed, can happen without patching Bitcoin Core source code itself.

This is not about debugging the start-up sequence specifically, but rather debugging *from an early stage*. Maybe one has set 5 breakpoints, some of which are hit during *bitcoind* init, and some which are hit by re-running a Python test which calls a set of RPCs.

fanquake commented:
Ideally we wouldn't be adding debug code, to production binaries, to facilitate using an IDE.

Can you configure your IDE to do something like the equivalent of using `gdb`/`lldb`, and `process attach --name bitcoind --waitfor`? So it just grabs bitcoind once it spins up?

hodlinator commented:
> Ideally we wouldn't be adding debug code, to production binaries, to facilitate using an IDE.

Agree it should not be in production binaries, should be `#ifdef`ed away. Added as another reason for draft status.
 
> Can you configure your IDE to do something like the equivalent of using `gdb`/`lldb`, and `process attach --name bitcoind --waitfor`? So it just grabs bitcoind once it spins up?

I've never seen something like that in an IDE/editor, but I'll have a look, might exist a debugger console which allows that kind of thing.

This PR does however allow for attaching to a specific node instance of bitcoind if a Python tests runs several nodes at once. (I'm considering changing the index to refer to how many bitcoind-launches one has done, some tests just relaunch node 0 a bunch of times, and one might be interested in debugging it on the 3rd execution).

i-am-yuvi commented:
Agree with @maflcko @fanquake, not sure if we want to add that to the production code.

hodlinator commented:
Added tentative `#ifdef` solution in latest push. CMake part could probably be improved since now I think the `#define` leaks into the CMake projects public interface, which should not be necessary.

hodlinator commented:
- Added CMake option `WAIT_FOR_DEBUGGER`.
- Changed the Python logic from debugging specific nodes, to debugging specific node runs/executions, which is more specific.
- Use `prctl()` on Linux to allow any process to attach to `bitcoind`. This removes the necessity to disable `kernel.yama.ptrace_scope` on some systems, which may be seen as decreasing security, but is compiled out unless `WAIT_FOR_DEBUGGER=ON`.
- (Confirmed PR works on Windows).
- (Rebased).

hodlinator commented:
Realized one could start a debugger session from the CLI:
`$ subl --command "debugger {\"action\": \"start\", \"configuration\": {\"name\": \"foo\", \"type\": \"lldb\", \"request\": \"launch\", \"program\": \"\${project_path}/build/bin/bitcoind\", \"args\": [\"-noconnect\"]}}"`
or
`$ kgx -- $(which lldb) -o run -- /home/hodlinator/bc/3/build/bin/bitcoind -noconnect`

So I've explored a variation of this which does not change the C++ side at all - https://github.com/hodlinator/bitcoin/tree/refs/heads/2025/09/launch_debugger. The approach is to launch the debugger from inside `TestNode` with the bitcoind command line as parameters. A stumbling block was that both debugger commands above exit immediately instead of running until bitcoind exits. Both document supporting `--wait` but Sublime Text requires passing in a file to be opened (`--wait dummyfile`) and the kgx Gnome Console has it marked as "TODO".

This lead me to create the `launch_and_wait.py` shim which simulates bitcoind process lifetime behavior. The shim has the following drawbacks:
* Requires manually sending a HTTP GET request for it to exit. - Maybe there's some way for it to monitor PID-files and exit automatically based off that for most tests which don't mess around with that file.
* While it could pass on environment variables (or in the case of Sublime Text have them specified through `configuration.env`), I don't think it's possible to pass through stderr/stdout since the debug launch commands exit quickly. - Maybe there's some way for it to monitor the node's stdout/stderr files and copy their contents to its own stdout/stderr, but it seems a bit messy.

Attaching to the process seems less error prone. After having worked on the above I realized attaching the debugger could be automated, so added that as a second commit in this PR.

If reviewers insist I could continue exploring the launch_debugger route to see how monitoring PID-files (or just require something like `--wait dummyfile`) and replicating stdout/stderr contents pans out.

l0rinc commented on src/bitcoind.cpp:
```diff
@@ -254,6 +320,10 @@ static bool AppInit(NodeContext& node)
 MAIN_FUNCTION
 {
+#ifdef WAIT_FOR_DEBUGGER
+    if (int ret{HandleWaitForDebugger(argc, argv)}; ret != EXIT_SUCCESS) return ret;
```
let's not propagate these further:
```suggestion
    if (int ret{HandleWaitForDebugger(std::span{argv + 1, size_t(argc - 1)})}; ret != EXIT_SUCCESS) return ret;
```

l0rinc commented on CMakeLists.txt:
```diff
@@ -120,6 +120,7 @@ cmake_dependent_option(BUILD_WALLET_TOOL "Build bitcoin-wallet tool." ${BUILD_TE
 option(REDUCE_EXPORTS "Attempt to reduce exported symbols in the resulting executables." OFF)
 option(WERROR "Treat compiler warnings as errors." OFF)
 option(WITH_CCACHE "Attempt to use ccache for compiling." ON)
+option(WAIT_FOR_DEBUGGER "Support for waiting during startup for debugger to be attached." OFF)
```
```suggestion
option(WAIT_FOR_DEBUGGER "Pause at startup until a debugger attaches (-waitfordebugger)." OFF)

```

l0rinc reviewed (commented):
Concept ACK, I'm always struggling with debugging these tests with multiple nodes, I usually end up just printing their internal states.

Thanks for working on this. I went through the code, tried to find simpler solutions to minimize and modernize the code we need to add to production code. I didn't get to trying this out yet, so I'm not sure my suggestions make sense, I just checked if it compiles on a Mac or Linux. Will try to test it properly next week.

Since we're editing production code for this I would prefer having a test which checks that `support was not included as WAIT_FOR_DEBUGGER` is logged, proving that we're correctly excluding it from code (it's fine if the test isn't passing in debug mode)

hodlinator reviewed (commented):
Thanks for reviewing @l0rinc!

> Since we're editing production code for this I would prefer having a test which checks that `support was not included as WAIT_FOR_DEBUGGER` is logged, proving that we're correctly excluding it from code (it's fine if the test isn't passing in debug mode)

Feel it would be a bit performative since `#ifdef`s and default-`OFF` CMake variables are pretty cut & dry. Maybe there's some scenario I haven't thought of.

l0rinc commented on src/init.cpp:
```diff
@@ -1125,6 +1126,12 @@ bool AppInitParameterInteraction(const ArgsManager& args)
 
+#ifndef WAIT_FOR_DEBUGGER
+    if (args.GetBoolArg("-waitfordebugger", false)) {
+        return InitError(Untranslated("-waitfordebugger support was not included as WAIT_FOR_DEBUGGER is not #defined."));
```
> add a functional test that runs bitcoind to confirm that `-waitfordebugger` triggers this error message

Yes, it would be reassuring to me that it's not on by default.

l0rinc commented on src/bitcoind.cpp:
```diff
@@ -159,6 +171,60 @@ static bool ProcessInitCommands(ArgsManager& args)
+
+                    std::getline(sf, s);
+                }
+                attached = pid > 0;
```
Ah, so it's not the negative that we're guarding, in that case maybe:
```suggestion
                attached = pid != 0;
```

l0rinc commented on src/bitcoind.cpp:
```diff
@@ -159,6 +171,60 @@ static bool ProcessInitCommands(ArgsManager& args)
+                    }
+
+                    std::getline(sf, s);
+                }
```
Precompiled regexes are optimized to NFAs, but this one doesn't even backtrack so it's likely really fast and very easy to read.

But looking at https://github.com/catchorg/Catch2/blob/devel/src/catch2/internal/catch_debugger.cpp#L95 it seems to me that we don't even need to parse it, just check if it starts with zero or not, maybe something like:
```C++
for (std::string line; !attached && std::getline(status, line);) {
    if (line.starts_with("TracerPid:")) {
        attached = line.find_first_of("123456789") != std::string::npos;
    }
}
```

hodlinator reviewed (commented):
Latest push adds a functional test and breaks apart commits somewhat.

hodlinator committed:
qa: Add wait_for_debugger param for node (re)starts

Enables tweaking functional tests to start debugging a specific node run/execution.

l0rinc commented on test/functional/feature_debugging.py:
```diff
@@ -0,0 +1,129 @@
+#!/usr/bin/env python3
+# Copyright (c) 2019-2021 The Bitcoin Core developers
```
```suggestion
# Copyright (c) 2019-present The Bitcoin Core developers
```

hodlinator commented:
Latest push adds 9d1de2e8b3c73edc44bb3f7578ec96c2b91c2dd8 to support modifying a functional test to make a specific node (re)start trigger attaching of a debugger. This may be a more natural alternative when humans read through functional test code and want to select a specific node instance to debug.

(It also removes a header which caused a build failure on Alpine/musl, and adds helpful details in 2 commit messages).

There has been some feedback to see if we could avoid changing the functional test framework and move most of this feature to */contrib/*, so I might try to figure out alternative approaches.

hodlinator committed:
qa: Add --debug_runs

Useful for debugging one or several bitcoind nodes running in the context of a functional test.

You can run a test with -ldebug once first to get output with run#s to see which you are interested in.

hodlinator committed:
qa: Add --debug_cmd

Enables automated attaching of debuggers.

The special identifier "$PID$" is substituted by the actual bitcoind process id number.

Attaching with LLDB in Tmux pane:
$ ./build/test/functional/feature_loadblock.py --debug_runs 2 --debug_cmd "tmux split-window -h lldb -p \$PID\$"

Attaching with Sublime Text Debugger:
$ ./build/test/functional/feature_loadblock.py --debug_runs 2 --debug_cmd "subl --command \"debugger {\\\"action\\\": \\\"start\\\", \\\"configuration\\\": {\\\"name\\\": \\\"foo\\\", \\\"type\\\": \\\"lldb\\\", \\\"request\\\": \\\"attach\\\", \\\"program\\\": \\\"dummyprogram\\\", \\\"pid\\\": \\\"\$PID\$\\\"}}\""

hodlinator commented:
Changes compared to f352bd9d246b20579b9b4ccec366bb694df5e051:
* feature_debugging.py - Added stderr output upon failure when we expect to be able to attach. Switched to `subprocess.run(..., text=True)` in spirit of fab085c15f7221986f73af7e05e799edf3eadaf0.
* Renamed CMake option `WAIT_FOR_DEBUGGER` -> `ENABLE_WAIT_FOR_DEBUGGER` to align convention with other options (also renamed `constexpr bool` in bitcoind.cpp from `WAIT_FOR_DEBUGGER_ENABLE` -> `WAIT_FOR_DEBUGGER_SUPPORT` to disambiguate).
* Added config.ini support for `ENABLE_WAIT_FOR_DEBUGGER` and implemented `is_wait_for_debugger_compiled()` instead of custom parsing of CMakeCache.txt in feature_debugging.py.
* Added comment in test_node.py explaining that functional test will continue executing even though a debugger has not been attached yet.
* Switched from mentioning Gnome Console in commit message and developer-notes.md to using Tmux as an example.
* Switched example test from feature_abortnode.py to feature_loadblock.py since the former causes a node to crash.
* Made `--debug_cmd` without `--debug_runs` not enable it for all runs, but instead require modifying test code to specify `wait_for_debugger=True` or use `--debug_runs`.

l0rinc commented on test/functional/feature_debugging.py:
```diff
@@ -0,0 +1,115 @@
+
+        assert os.path.exists(f'/proc/{node_pid}')
+
+        # TODO: When wallets are loaded in bitcoind, we should ideally disable:
```
who is this TODO meant for?

hodlinator replied:
Sorry for leaving this around...

I found that we could add this somewhere:
```C++
#if defined(__linux__)
#include <sys/prctl.h>
#endif
...
#if defined(__linux__)
    if (int ret{prctl(PR_SET_DUMPABLE, 0)}; ret != EXIT_SUCCESS) {
        return ret;
    }
#endif
```
Which both prevents reading from memory:
```python
assert_raises(PermissionError, lambda: open(f'/proc/{node_pid}/mem', 'rb').seek(1))
```
and disables core dump generation when killing the process with SIGABRT:
```
pkill bitcoind -SIGABRT
```
But I think that's an idea for a later, wallet-specific PR. I'll scrub it from this one.

l0rinc commented on src/bitcoind.cpp:
```diff
@@ -164,6 +183,59 @@ static bool ProcessInitCommands(interfaces::Init& init, ArgsManager& args)
+            ret != EXIT_SUCCESS) {
+            fputs("Could not detect debugger, sysctl() for fetching traced-status failed.\n", stderr);
+            return ret;
+        }
```
We could zero-init `info` and return `EXIT_FAILURE`:
```suggestion
        kinfo_proc info{};
        size_t info_size{sizeof(info)};
        if (sysctl(/*name=*/mib.data(), /*namelen=*/mib.size(), /*oldp=*/&info, /*oldlenp=*/&info_size, /*newp=*/nullptr, /*newlen=*/0)) {
            fputs("Could not detect debugger, sysctl() for fetching traced-status failed.\n", stderr);
            return EXIT_FAILURE;
        }
```

hodlinator commented on test/functional/test_framework/test_node.py:
Oof. You're talking about this part of CPython? https://github.com/python/cpython/blob/v3.14.0a1/Lib/subprocess.py#L1996-L2008

I'm not sure we can allow returncode 0 under normal operation... I guess maybe we could double-check some other way whether the process is alive.. but what to do with the now bogus `self.process.returncode`.. setting it back to `None` from the outside seems flaky.

l0rinc replied:
Is the above diff of `if self.process.poll() is not None and self.process.returncode != 0:` unacceptable? What's the drawback of testing the exit code of the polled process?

hodlinator reviewed (commented):
Rebased and responded to input from https://github.com/bitcoin/bitcoin/pull/31723#pullrequestreview-3372365346 + removed explicit UTF-8 along the lines of #33702.

hodlinator committed:
node: Add -waitfordebugger

Makes bitcoind spin during startup, waiting for a debugger to be attached. Once a debugger is detected, execution will continue.

hodlinator committed:
qa: Add feature_debugging.py and feature_debugging_disallowed.py

feature_debugging.py verifies that debugging works.

feature_debugging_disallowed.py verifies that debugging bitcoind is not allowed by default on Linux, at least in sane environments.

hodlinator reviewed (commented):
In the latest push ([compare](https://github.com/bitcoin/bitcoin/compare/22b6f3fa2e73af0922734a278777d3271fe0f3d4..254091d9151b4bb389e6db866d4dd60d39f53903)), I've broken out test/functional/feature_debugging_disallowed.py from test/functional/feature_debugging.py.
* feature_debugging_disallowed.py only works on Linux.
* feature_debugging.py should work on all platforms. It has passed MacOS CI (https://github.com/hodlinator/bitcoin/actions/runs/21515676070/job/61993329472#step:9:3358) but I suspect it would fail if one turns on `ENABLE_WAIT_FOR_DEBUGGER` as described at https://github.com/bitcoin/bitcoin/pull/31723#discussion_r2698217840.

Also improved the GDB/LLDB scripts in feature_debugging.py by adding a second `continue` command, which will only execute if the first continue stops at a breakpoint. This means that if the breakpoint isn't hit, the tests will timeout.

Noticed that lldb.exe was failing for feature_debugging.py on CI and in my local Windows environment. When executing it manually, it complains about a missing liblldb.dll, see https://stackoverflow.com/questions/76520754/visual-studio-lldb-exe-liblldb-dll-not-found. So added code to check for lldb not being operational.

(Edit: The following push of a2e01576c37a6f1aca0db42184f6632f9c91543e just removes a logging statement which tried to log before logging was available).

instagibbs commented:
two points:

1. I don't think this enables much. One can pdb.set_trace to stop the test wherever we want, then `gdb -p <pid>` to the process, continue gdb or whatever, then continue the test. It's a bit of friction, but not that much. I'd suggest trying that out first and see if that doesn't suffice for people's use cases.

2. It *would* be useful if we were in a situation with many processes running around, as this feature is in CLN, and helped a lot with sub binaries spawning off, doing stuff without any other interaction. I don't think Core has this issue(? yet?)

hodlinator commented:
I appreciate the feedback @instagibbs! I've switched to inserting
```
	import pdb
    pdb.set_trace()
```
and manually attaching for now. Will reopen this if/when I find/rediscover a concrete example use-case for it (like needing to attach early).

Closing for now.

instagibbs commented:
I've seen *very* seasoned Core devs not know about this debugging pattern, if we can surface it more easily in docs we should.
